File: 1_Spiral_Traversal/spiral_matrix.py
def spirally_traverse(matrix):
    top, left = 0, 0
    bottom, right = len(matrix)-1, len(matrix[0])-1
    result = []
    while top <= bottom and left <= right:
        for i in range(left, right+1):
            result.append(matrix[top][i])
        top += 1
        for i in range(top, bottom+1):
            result.append(matrix[i][right])
        right -= 1
        if top <= bottom:
            for i in range(right, left-1, -1):
                result.append(matrix[bottom][i])
            bottom -= 1
        if left <= right:
            for i in range(bottom, top-1, -1):
                result.append(matrix[i][left])
            left += 1
    return result

r, c = map(int, input().split())
matrix = [list(map(int, input().split())) for _ in range(r)]
print(*spirally_traverse(matrix))

File: 1_Spiral_Traversal/spiral_matrix.c
#include <stdio.h>
void spirallyTraverse(int r, int c, int matrix[100][100]) {
    int top = 0, bottom = r - 1, left = 0, right = c - 1;
    while (top <= bottom && left <= right) {
        for (int i = left; i <= right; i++) printf("%d ", matrix[top][i]);
        top++;
        for (int i = top; i <= bottom; i++) printf("%d ", matrix[i][right]);
        right--;
        if (top <= bottom) {
            for (int i = right; i >= left; i--) printf("%d ", matrix[bottom][i]);
            bottom--;
        }
        if (left <= right) {
            for (int i = bottom; i >= top; i--) printf("%d ", matrix[i][left]);
            left++;
        }
    }
}
int main() {
    int r, c;
    scanf("%d %d", &r, &c);
    int matrix[100][100];
    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++)
            scanf("%d", &matrix[i][j]);
    spirallyTraverse(r, c, matrix);
    return 0;
}
